### PR TITLE
Events

### PR DESCRIPTION
**Pull request by Anze1m**
# Checklist:

- [x] Code follows the style guidelines
- [x] Documentation updated
- [x] Corner cases evaluated and handled
- [x] index.js exports updated
- [x] Automatic tests added/updated
- [x] Automatic tests pass
- [ ] Sanity check passed
- [x] Self-review passed
- [x] Version updated

### CODE REVIEW

**Comment by vanruch** [comment]
I'm missing changes in:
- ChallengesWrapper:
    - challenges
    - resolvedChallenges
    - timedOutChallenges
- UploadsWrapper:
    - getUploadData

File: contracts/Storage/ChallengesEventEmitter.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    function challengeResolved(address sheltererId, bytes32 bundleId, bytes32 challengeId, address resolverId) public onlyContextInternalCalls {
+        emit ChallengeResolved(sheltererId, bundleId, challengeId, resolverId);
+    }
+
+    function challengeTimeout(address sheltererId, bytes32 bundleId, bytes32 challengeId, uint penalty) public onlyContextInternalCalls {
+        emit ChallengeTimeout(sheltererId, bundleId, challengeId, penalty);
+    }
+}
```
Comment (vanruch) [inline contracts/Storage/ChallengesEventEmitter.sol]: I would propose a naming convention in event emitters to add on before an event name. e.g. `onChallengeTimeout`

File: contracts/Storage/RolesEventEmitter.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import "../Boilerplate/Head.sol";
+import "../Configuration/Consts.sol";
+
+
+contract RolesEventEmitter is Base {
+
+    constructor(Head _head) Base(_head) public {}
+
+    event MasternodeOnboarded(address nodeAddress, uint placedDeposit, string nodeUrl, Consts.NodeType role);
```
Comment (vanruch) [inline contracts/Storage/RolesEventEmitter.sol]: We don't use the term 'masternode' anywhere but here iirc, mb just leave node?

File: test/contracts/front/challenges.js [hunk 3]
```diff
@@ -292,7 +300,7 @@ describe('Challenges Contract', () => {
     });
 
     it('Fails if added same challenge twice', async () => {
-      expect(await startChallenge(shelterer, bundleId, challenger, userChallengeFee)).to.emitEvent('ChallengeCreated');
+      expect(startChallenge(shelterer, bundleId, challenger, userChallengeFee)).to.be.eventually.fulfilled;
```
Comment (vanruch) [inline test/contracts/front/challenges.js]: missing await

File: test/helpers/web3EventObserver.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/*
+Copyright: Ambrosus Technologies GmbH
+Email: tech@ambrosus.com
+
+This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
+*/
+
+const observeEventEmission = async (web3, codeBlock, contract, eventName) => {
```
Comment (vanruch) [inline test/helpers/web3EventObserver.js]: Don't like the word observe, for me it implies some kind of subscription or something

Comment (Anze1m) [reply test/helpers/web3EventObserver.js]: This makes it consistent with `web3BalanceObserver` and `observeBalanceChange`.
If you have better idea, then plz come up with one.

File: contracts/Storage/ChallengesEventEmitter.sol (hunk 1 above)
Comment (Anze1m) [reply contracts/Storage/ChallengesEventEmitter.sol]: I think that function which only emit events should be named the same as these events. Imo the most intuitive solution.

File: src/actions/upload_actions.js [hunk 5]
```diff
@@ -47,7 +48,8 @@ export default class UploadActions {
       return null;
     }
     const timestamp = await this.blockchainStateWrapper.getBlockTimestamp(uploadBlock);
-    const {transactionHash} = await this.uploadsWrapper.getUploadData(bundleId, uploadBlock);
+    const events = await this.challengesEventEmitterWrapper.challenges(uploadBlock, uploadBlock);
+    const [{transactionHash}] = events.filter(({returnedValues}) => returnedValues.bundleId === bundleId);
```
Comment (spherefoundry) [inline src/actions/upload_actions.js]: You could use here `find` instead of `filter` which would stop on the first match.

File: src/tasks/run.js [hunk 6]
```diff
@@ -39,6 +39,7 @@ import PayoutsWrapper from '../wrappers/payouts_wrapper';
 import ChallengeActions from '../actions/challenge_actions';
 import ChallengeWrapper from '../wrappers/challenge_wrapper';
 import BlockchainStateWrapper from '../wrappers/blockchain_state_wrapper';
+import ChallengesEventEmitter from '../wrappers/challenges_event_emitter_wrapper';
```
Comment (spherefoundry) [inline src/tasks/run.js]: I'm missing the `*Wrapper` suffix here

File: test/contracts/front/challenges.js [hunk 7]
```diff
@@ -189,7 +195,7 @@ describe('Challenges Contract', () => {
     });
 
     it('Fails if added same challenge twice', async () => {
-      await expect(startChallengeForSystem(uploader, bundleId, SYSTEM_CHALLENGES_COUNT, context, systemChallengeFee)).to.be.eventually.fulfilled;
+      expect(startChallengeForSystem(uploader, bundleId, SYSTEM_CHALLENGES_COUNT, context, systemChallengeFee)).to.be.eventually.fulfilled;
```
Comment (spherefoundry) [inline test/contracts/front/challenges.js]: but why?

File: test/contracts/front/challenges.js [hunk 8]
```diff
@@ -379,6 +387,14 @@ describe('Challenges Contract', () => {
       cooldown = parseInt(await getCooldown(), 10);
     });
 
+    it('Emits proper event', async () => {
+      const events = await observeEventEmission(web3, () => resolveChallenge(challengeId, resolver), challengesEventEmitter, 'ChallengeResolved');
+      expect(events.length).to.equal(1);
```
Comment (spherefoundry) [inline test/contracts/front/challenges.js]: A very similar pattern is repeated some 3 times in this file. Maybe a helper function?

File: contracts/Storage/ChallengesEventEmitter.sol (hunk 1 above)
Comment (vanruch) [reply contracts/Storage/ChallengesEventEmitter.sol]: I think it should be name so it's clear that it only emits event and does nothing more

File: test/helpers/web3EventObserver.js (hunk 4 above)
Comment (spherefoundry) [reply test/helpers/web3EventObserver.js]: `captureEventEmission`?

File: test/helpers/web3EventObserver.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  await codeBlock();
+  const toBlock = await web3.eth.getBlockNumber();
+  return contract.getPastEvents(eventName, {fromBlock, toBlock});
+};
+
+const expectEventEmission = async (web3, codeBlock, contract, eventName, eventArguments) => {
+  const events = await observeEventEmission(web3, codeBlock, contract, eventName);
+  expect(events.length).to.equal(1);
+  Object.keys(eventArguments).forEach((eventArgumentName) => {
+    expect(events[0].returnValues[`${eventArgumentName}`]).to.equal(eventArguments[`${eventArgumentName}`]);
```
Comment (spherefoundry) [inline test/helpers/web3EventObserver.js]: `${eventArgumentName}` === eventArgumentName ?
